### PR TITLE
Update AUI to respond to SysColourChanged events and update the art a…

### PR DESCRIPTION
**Pull request by dkulp**
…ccordingly (on OSX)

Add a method to the various AUI art providers to have them recalc their colors on a SysColourChanged event.   Make sure it's called from the various AUI classes.

### CODE REVIEW

**Comment by dkulp** [comment]
Note that this is a start.... toolbars are working well, other parts aren't yet.

**Comment by dkulp** [comment]
Pushed an update... the parts of AUI that my app use are now flipping back and forth between Light/Dark fairly well.

File: src/propgrid/editors.cpp [hunk 1]
```diff
@@ -742,6 +742,8 @@ void wxPropertyGrid::OnComboItemPaint( const wxPGComboBox* pCb,
                                        int flags )
 {
     wxPGProperty* p = GetSelection();
+    if (!p) return;
```
Comment (vadz) [inline src/propgrid/editors.cpp]: This seems unrelated, so it'd be be better to extract this change in a separate commit and explain when exactly does this problem happen in the commit message.

File: src/common/wincmn.cpp [hunk 2]
```diff
@@ -2961,10 +2961,10 @@ void wxWindowBase::OnSysColourChanged(wxSysColourChangedEvent& WXUNUSED(event))
             event2.SetEventObject(win);
             win->GetEventHandler()->ProcessEvent(event2);
         }
-
         node = node->GetNext();
     }
 
+    event.Skip(true);
```
Comment (vadz) [inline src/common/wincmn.cpp]: Why do we have to do this now? Is there some default handling of this event at the underlying toolkit level? If so, this probably ought to be emphasized in the documentation (unless we prefer to always forward this event to the native handler).

File: src/aui/tabart.cpp [hunk 3]
```diff
@@ -164,12 +164,35 @@ wxAuiGenericTabArt::wxAuiGenericTabArt()
...
+
+void wxAuiGenericTabArt::SysColourChanged()
+{
 #if defined( __WXMAC__ ) && wxOSX_USE_COCOA_OR_CARBON
     wxColor baseColour = wxSystemSettings::GetColour(wxSYS_COLOUR_WINDOW);
 #else
     wxColor baseColour = wxSystemSettings::GetColour(wxSYS_COLOUR_3DFACE);
 #endif
-
+
```
Comment (vadz) [inline src/aui/tabart.cpp]: Please try to avoid adding extra whitespace, it's minor but annoying.

Putting `exec git diff-index --check --cached HEAD --` in your `.git/hooks/pre-commit` and doing `git config core.whitespace blank-at-eol,space-before-tab,cr-at-eol` will prevent you from committing such whitespace additions accidentally.

File: src/aui/framemanager.cpp [hunk 4]
```diff
@@ -647,7 +648,11 @@ wxAuiManager::~wxAuiManager()
 
     delete m_art;
 }
-
+void wxAuiManager::OnSysColourChanged(wxSysColourChangedEvent& event) {
```
Comment (vadz) [inline src/aui/framemanager.cpp]: Minor style: please separate functions with blank lines and put `{` on its own line for consistency.

File: src/aui/framemanager.cpp [hunk 5]
```diff
@@ -647,7 +648,11 @@ wxAuiManager::~wxAuiManager()
...
+    m_art->SysColourChanged();
+    m_frame->Refresh();
```
Comment (vadz) [inline src/aui/framemanager.cpp]: It seems a bit weird to have to refresh the frame from this handler. Shouldn't it be refreshed from its own handler?

File: include/wx/aui/auibar.h [hunk 6]
```diff
@@ -343,6 +343,10 @@ class WXDLLIMPEXP_AUI wxAuiToolBarArt
     virtual int ShowDropDown(
                          wxWindow* wnd,
                          const wxAuiToolBarItemArray& items) = 0;
+    
+    //Provide opportunity for suclasses to recalculate colors
+    virtual void SysColourChanged() {}
```
Comment (vadz) [inline include/wx/aui/auibar.h]: I don't really like this name, it seems weird to call this function from the ctor for example (because nothing has actually changed).

What about `SetColoursFromSystemTheme()` or maybe `UpdateColoursFromSystem()` or something like that?

**Review by vadz — COMMENTED** [review]
Thanks for the fixes!

I definitely agree that we need to do something like this, but I wonder about the need to skip the event in all handlers and I'd also like to find a better name for `SysColourChanged()` if possible.

File: src/propgrid/editors.cpp (hunk 1 above)
Comment (dkulp) [reply src/propgrid/editors.cpp]: Submitted https://github.com/wxWidgets/wxWidgets/pull/1008

File: src/aui/framemanager.cpp (hunk 4 above)
Comment (dkulp) [reply src/aui/framemanager.cpp]: Will fix...

File: src/aui/framemanager.cpp (hunk 5 above)
Comment (dkulp) [reply src/aui/framemanager.cpp]: I had a strange race condition where the frame would gets it's wxSysColourChangedEvent and refresh itself prior to the AuiManager getting its event so it wasn't picking up the new colors. That said, I cannot reproduce it right now.

File: include/wx/aui/auibar.h (hunk 6 above)
Comment (dkulp) [reply include/wx/aui/auibar.h]: I'll update with UpdateColoursFromSystem()

File: src/aui/framemanager.cpp (hunk 5 above)
Comment (dkulp) [reply src/aui/framemanager.cpp]: No, the m_frame needs to be fully refreshed.   Otherwise none of the docked panels get updated and they continue to display their old colors.  The other option might be to go and find all the docked panels and windows and refresh each one individually, but the full refresh seems to work best.

**Comment by vadz** [comment]
Thanks! Applied with minor changes now.
